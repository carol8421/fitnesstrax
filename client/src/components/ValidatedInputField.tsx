import { Option } from "ld-ambiguity"
import React from "react"

import * as csn from "../classnames"
import InputField from "./InputField"

//import { maybe, isSomething } from "../common"

/* Required parameters:
 *
 * value: the data value to be rendered
 * render: (value) => str
 * parser: (str) => maybe value
 *
 * Optional parameter:
 *
 * onChange: (value) => action to take when a valid value is in the field
 */

export interface Props<A> {
  value: Option<A>
  placeholder: string
  render: (_: A) => string
  parse: (_: string) => Option<A>
  onChange: ((_: A) => void)
}

interface State<A> {
  text: Option<string>
  value: Option<A>
}

export class ValidatedInputField<A> extends React.Component<
  Props<A>,
  State<A>
> {
  state: State<A>

  constructor(props: Props<A>) {
    super(props)

    this.state = {
      text: props.value.map(props.render),
      value: props.value,
    }
  }

  update(value: string) {
    const { parse, onChange } = this.props
    const res = parse(value)
    this.setState({ text: Option.Some(value), value: res })
    res.map(v => onChange(v))
  }

  render() {
    const { placeholder } = this.props
    const { value } = this.state
    const valid = value
      .map(_ => ({
        valid: true,
      }))
      .or({ valid: false })
    return (
      <span className={csn.classnames(valid)}>
        <InputField
          classNames={valid}
          value={this.state.text}
          onChange={value => this.update(value)}
          placeholder={placeholder}
        />
      </span>
    )
  }
}

export default ValidatedInputField
