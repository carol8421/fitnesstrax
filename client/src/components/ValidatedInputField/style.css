input {
  margin: 0.25em;
}

input.invalid {
  border: 2px solid rgb(255, 96, 96);
  padding: 2px;
  border-radius: 2px;
}
