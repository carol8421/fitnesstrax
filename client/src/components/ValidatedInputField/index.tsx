import { Option, Result } from "ld-ambiguity"
import React from "react"

import * as csn from "../../classnames"
import InputField from "../InputField"
import "./style.css"

/* Required parameters:
 *
 * value: the data value to be rendered
 * render: (value) => str
 * parser: (str) => maybe value
 *
 * Optional parameter:
 *
 * onChange: (value) => action to take when a valid value is in the field
 */

export interface Props<A> {
  value: Option<A>
  placeholder: string
  render: (_: A) => string
  parse: (_: string) => Result<Option<A>, string>
  onChange: ((_: Option<A>) => void)
}

interface State<A> {
  text: string
  value: Option<A>
  valid: boolean
}

export class ValidatedInputField<A> extends React.Component<
  Props<A>,
  State<A>
> {
  state: State<A>

  constructor(props: Props<A>) {
    super(props)

    this.state = {
      text: props.value.map(props.render).or(""),
      value: props.value,
      valid: true,
    }
  }

  update(value: string) {
    const { parse, onChange } = this.props
    const res = parse(value)
    if (res.isOk()) {
      this.setState({ text: value, value: res.unwrap(), valid: true })
      res.map(v => onChange(v))
    } else {
      this.setState({ text: value, value: Option.None(), valid: false })
    }
  }

  render() {
    const { placeholder } = this.props
    const { value, valid } = this.state
    const validityClass = valid
      ? { valid: true, invalid: false }
      : { valid: false, invalid: true }
    return (
      <div className={csn.classnames(validityClass)}>
        <InputField
          classNames={validityClass}
          value={this.state.text}
          onChange={value => this.update(value)}
          placeholder={placeholder}
        />
      </div>
    )
  }
}

export default ValidatedInputField
