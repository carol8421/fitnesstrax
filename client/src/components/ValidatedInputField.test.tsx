import React from "react"
import { mount } from "enzyme"

import Option from "../option"
import ValidatedInputField from "./ValidatedInputField"

describe("ValidatedInputField", () => {
  it("renders", () => {
    const onChange = jest.fn()
    const wrapper = mount(
      <ValidatedInputField
        value={Option.Some(15)}
        placeholder="validated input field"
        render={(val: number): string => val.toString()}
        parse={(str: string): Option<number> => Option.fromNaN(parseInt(str))}
        onChange={onChange}
      />,
    )
    expect(wrapper).toMatchSnapshot()
    expect(onChange).not.toHaveBeenCalled()
    expect(wrapper.find("input").prop("value")).toEqual("15")
  })

  it("handles valid updates", () => {
    const onChange = jest.fn()
    const wrapper = mount(
      <ValidatedInputField
        value={Option.Some(15)}
        placeholder="validated input field"
        render={(val: number): string => val.toString()}
        parse={(str: string): Option<number> => Option.fromNaN(parseInt(str))}
        onChange={onChange}
      />,
    )
    wrapper.find("input").simulate("change", { target: { value: "16" } })
    expect(wrapper).toMatchSnapshot()
    expect(onChange).toHaveBeenCalledWith(16)
    expect(wrapper.find("input").prop("value")).toEqual("16")
  })

  it("handles invalid updates", () => {
    const onChange = jest.fn()
    const wrapper = mount(
      <ValidatedInputField
        value={Option.Some(15)}
        placeholder="validated input field"
        render={(val: number): string => val.toString()}
        parse={(str: string): Option<number> => Option.fromNaN(parseInt(str))}
        onChange={onChange}
      />,
    )
    wrapper.find("input").simulate("change", { target: { value: "abcdefg" } })
    expect(wrapper).toMatchSnapshot()
    expect(onChange).not.toHaveBeenCalled()
    expect(wrapper.find("input").prop("value")).toEqual("abcdefg")
  })
})
