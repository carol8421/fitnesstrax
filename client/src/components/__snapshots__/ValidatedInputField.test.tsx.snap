// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatedInputField handles invalid updates 1`] = `
<ValidatedInputField
  onChange={[MockFunction]}
  parse={[Function]}
  placeholder="validated input field"
  render={[Function]}
  value={
    t {
      "val_": 15,
    }
  }
>
  <span
    className=""
  >
    <InputField
      classNames={
        Object {
          "valid": false,
        }
      }
      onChange={[Function]}
      placeholder="validated input field"
      value={
        t {
          "val_": "abcdefg",
        }
      }
    >
      <input
        className=""
        onChange={[Function]}
        placeholder="validated input field"
        type="text"
        value="abcdefg"
      />
    </InputField>
  </span>
</ValidatedInputField>
`;

exports[`ValidatedInputField handles valid updates 1`] = `
<ValidatedInputField
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          16,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  parse={[Function]}
  placeholder="validated input field"
  render={[Function]}
  value={
    t {
      "val_": 15,
    }
  }
>
  <span
    className="valid"
  >
    <InputField
      classNames={
        Object {
          "valid": true,
        }
      }
      onChange={[Function]}
      placeholder="validated input field"
      value={
        t {
          "val_": "16",
        }
      }
    >
      <input
        className="valid"
        onChange={[Function]}
        placeholder="validated input field"
        type="text"
        value="16"
      />
    </InputField>
  </span>
</ValidatedInputField>
`;

exports[`ValidatedInputField renders 1`] = `
<ValidatedInputField
  onChange={[MockFunction]}
  parse={[Function]}
  placeholder="validated input field"
  render={[Function]}
  value={
    t {
      "val_": 15,
    }
  }
>
  <span
    className="valid"
  >
    <InputField
      classNames={
        Object {
          "valid": true,
        }
      }
      onChange={[Function]}
      placeholder="validated input field"
      value={
        t {
          "val_": "15",
        }
      }
    >
      <input
        className="valid"
        onChange={[Function]}
        placeholder="validated input field"
        type="text"
        value="15"
      />
    </InputField>
  </span>
</ValidatedInputField>
`;
